---
title: "Long-read_analysis"
author: "Morgane Fierville & Ali Hamraoui"
date: "2024-09-06"
output:
  html_document:
    code_folding: show
    code_download: true
    toc: true
    toc_float: true
    number_sections: false
---
<style>
body {
text-align: justify}
</style>

<!-- Set default parameters for all chunks -->
```{r, setup, include = FALSE}
set.seed(1337L)
knitr::opts_chunk$set(echo = TRUE, # display code
                      # display chunk output
                      message = FALSE,
                      warning = FALSE,
                      fold_output = FALSE,
                      fold_plot = FALSE,
                      
                      # figure settings
                      fig.align = 'center',
                      fig.width = 20,
                      fig.height = 15)
```

This file is used to explore ways to visualize differentially expressed isoforms.

```{r library}
library(dplyr)
library(patchwork)
library(ggplot2)
library(biomaRt)
library(stringr)
library(reactable)
library(gprofiler2)
#.libPaths()
```

# Preparation

In this section, we set the global settings of the analysis. We will store data there :

```{r out_dir}
data_dir <- "./data/"
out_dir = "./output/"
```

We load the Seurat object :

```{r load_sobj}
seurat_obj = readRDS(paste0(out_dir, "/datasets/seurat_obj_annotated.rds"))
seurat_obj
```

This is the projection of interest :

```{r name2D}
name2D = "RNA_pca_10_tsne"
```

We load the DE isoforms table :

```{r load_switch_markers}
switch_markers = readRDS(paste0(out_dir, "/datasets/switch_markers_combined_all.rds"))
```

We set a custom functions to visualize expression levels for a given gene and isoform:

```{r see_isoforms_expression, class.source = "fold-hide"}
see_isoforms_expression = function(sobj, gene, limits) {
  isoforms = grep(paste0("^", gene, "\\."), rownames(sobj), value = TRUE)
  
  plot_list = lapply(isoforms, FUN = function(one_isoform) {
    p = Seurat::FeaturePlot(sobj, features = one_isoform,
                            reduction = name2D, order = TRUE) +
      ggplot2::labs(title = one_isoform) +
      ggplot2::scale_color_gradientn(limits = limits, colors = c("lightgray", "#FDBB84", "#EF6548", "#7F0000", "black")) +
      ggplot2::theme(aspect.ratio = 1,
                     plot.subtitle = element_text(hjust = 0.5)) +
      Seurat::NoAxes()
    return(p)
  })
  
  return(plot_list)
}

see_genes_expression <- function(gene){
  Seurat::FeaturePlot(seurat_obj, gene, reduction = name2D)+
      ggplot2::scale_color_gradientn(limits = c(0,4), colors = c("lightgray", "#FDBB84", "#EF6548", "#7F0000", "black")) +
      ggplot2::theme(aspect.ratio = 1,
                     plot.subtitle = element_text(hjust = 0.5)) +
      Seurat::NoAxes()
}
```

# Dataset

We visualize cells :

```{r see_cells, fig.height=6, fig.width=16, message=FALSE, warning=FALSE, paged.print=FALSE}
cell_type = Seurat::DimPlot(seurat_obj, reduction = name2D, group.by = "cluster_cell_type") +
  Seurat::NoAxes() +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_text(hjust = 0.5),
                 plot.subtitle = element_text(hjust = 0.5))

clusters = Seurat::DimPlot(seurat_obj, group.by = "seurat_clusters", label = TRUE,
                           reduction = name2D) +
  Seurat::NoAxes() + ggplot2::ggtitle("Clustering") +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_text(hjust = 0.5))

clusters | cell_type
```

We visualize the main gene markers:

```{r fig.height=6, fig.width=12}
see_genes_expression("H3f3b") | see_genes_expression("Sox11")
```

We visualize Cell cycle:

```{r fig.height=5, fig.width=5}
mmus_s = gorth(Seurat::cc.genes.updated.2019$s.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name
mmus_g2m = gorth(Seurat::cc.genes.updated.2019$g2m.genes, source_organism = "hsapiens", target_organism = "mmusculus")$ortholog_name

seurat_obj <- Seurat::CellCycleScoring(seurat_obj, assay="RNA", g2m.features = mmus_g2m, s.features = mmus_s)
Seurat::DimPlot(seurat_obj, reduction = name2D,
        group.by= "Phase")
```
# DE isoforms visualisation

In this section, we explore the `switch_marker` dataframe.

List of isoform switchs

```{r gene_switch_table, fig.height = 8, fig.width = 15}
isoswitch::gene_switch_table(switch_markers)
```

left: heatmap of number of unique genes per contrast between clusters.
right: volcano plot showing p-values and average logFC for each gene with an isoform switch.

```{r plot_marker_score, fig.height=6, fig.width=12}
pl1 <- isoswitch::plot_marker_matrix(seurat_obj, switch_markers) 
pl2 <- isoswitch::plot_marker_score(seurat_obj, switch_markers, facet=FALSE, overlaps=16)
pl1 | pl2 
```
individual volcano plots for each cluster.
 
```{r fig.height=12, fig.width=14}
isoswitch::plot_marker_score(seurat_obj, switch_markers, facet=TRUE, ncol=3)
```

## Heatmap

We represent a heatmap for top DE isoforms (log2FC>2 and P_val<0.005):

```{r top50, fig.height = 5, fig.width = 10}
top = switch_markers %>%
  group_by(cluster) %>%
  dplyr::filter(abs(p_val) < 0.05) %>%
  dplyr::filter(abs(avg_log2FC) > 2) %>%
  dplyr::filter(pct.1 > 0.1 | pct.2 > 0.1) %>%
  ungroup()

Seurat::DoHeatmap(seurat_obj, features = sort(top$feature[0:10]), assay = "ISO") 
```
## Specific genes

We visualize isoforms expression levels for specific genes. First, we set the assay to "ISO" because all expression levels are stored there.

```{r set_default_assay}
Seurat::DefaultAssay(seurat_obj) = "ISO"
```

For Cdkn2c:
```{r see_cdkn2c, fig.width = 12, fig.height = 6}
see_isoforms_expression(seurat_obj, "Clta", c(0,4)) %>%
  patchwork::wrap_plots(., ncol = 3)
```

```{r fig.height=4, fig.width=16}
Seurat::FeaturePlot(seurat_obj, reduction = name2D, features = c("Clta..ENSMUST00000107849.9", "Clta..ENSMUST00000170241.7"), blend = TRUE, order = TRUE)
```

## DotPlot

```{r fig.height=6, fig.width=12}
features <- c("Clta", "Myl6", "Tecr", "Pkm", "Cdc42", "Smarcb1", "Rps24", "Zcrb1", "Gnas", "Zcrb1")
Seurat::DotPlot(seurat_obj, features = sort(unique(switch_markers[switch_markers$geneId %in% features, "feature"])) ) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```
## Report
In this section, we generate a small report.

We load annotation files:
```{r fig.height=6, fig.width=12}
gtf_df <- readRDS(paste0(data_dir,"annotation/gtf_rds.rds"))
gene_metadata <- readRDS(paste0(data_dir,"annotation/gene_metadata.rds"))
transcript_metadata <- readRDS(paste0(data_dir,"annotation/transcript_metadata.rds"))
```

For Clta:
```{r fig.height=6, fig.width=12}
gene_switches <- isoswitch::compute_switches(switch_markers, gene="Clta")
knitr::kable(isoswitch::format_switch_table(gene_switches))
#isoswitch::isoswitch_report(seurat_obj, "ISO", gene="Clta", marker_list=switch_markers, gtf_df, transcript_metadata) 
```

# R Session

```{r sessioninfo, echo = FALSE, fold_output = TRUE}
sessionInfo()
```
